transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,837 @@
+        }
+    }
+
+    #[derive(Debug, Clone)]
+    pub(crate) struct EndpointRef {
+        pub(super) reference: Arc<EndpointData>,
+        // MUST COME LATER so that the endpoint driver gets notified *after* its reference count is
+        // dropped.
+        pub(super) channel: Channel,
+    }
+}
+
+pub(crate) use channel_ref::EndpointRef;
+
+/// A QUIC endpoint. Each endpoint has its own configuration and listening socket.
+///
+/// You generally need at most one of these per thread. Endpoints are [`Send`] and [`Sync`], so you
+/// can share them among as many threads as you like. However, performance may be better if you
+/// have one per CPU core, as this reduces lock contention. Most applications will not need to
+/// worry about this. [`Endpoint`] tries to use fine-grained locking to reduce the overhead.
+///
+/// [`Endpoint`] wraps the underlying data structure in an [`Arc`], so cloning it just bumps the
+/// reference count. All state is shared between the clones. For example, you can pass different
+/// clones to [`<Endpoint as Transport>::listen_on`]. Each incoming connection will be received by
+/// exactly one of them.
+///
+/// The **only** valid [`Multiaddr`] to pass to [`<Endpoint as Transport>::listen_on`] is the one
+/// used to create the [`Endpoint`]. If you pass a different one, you will get
+/// [`TransportError::MultiaddrNotSupported`].
+#[derive(Debug, Clone)]
+pub struct Endpoint(EndpointRef);
+
+/// A handle that can be used to wait for the endpoint driver to finish.
+pub type JoinHandle = async_std::task::JoinHandle<Result<(), Error>>;
+
+#[derive(Debug)]
+pub(crate) struct SendToken(());
+
+impl Endpoint {
+    /// Construct a [`Endpoint`] with the given `Config`.
+    pub fn new(
+        config: Config,
+    ) -> Result<(Self, JoinHandle), TransportError<<Self as Transport>::Error>> {
+        let Config {
+            mut multiaddr,
+            client_config,
+            server_config,
+            endpoint_config,
+        } = config;
+        let socket_addr = if let Ok(sa) = multiaddr_to_socketaddr(&multiaddr) {
+            sa
+        } else {
+            return Err(TransportError::MultiaddrNotSupported(multiaddr));
+        };
+        // NOT blocking, as per man:bind(2), as we pass an IP address.
+        let socket = std::net::UdpSocket::bind(&socket_addr)
+            .map_err(|e| TransportError::Other(Error::IO(e)))?;
+        let port_is_zero = socket_addr.port() == 0;
+        let socket_addr = socket.local_addr().expect("this socket is bound; qed");
+        info!("bound socket to {:?}", socket_addr);
+        if port_is_zero {
+            assert_ne!(socket_addr.port(), 0);
+            assert_eq!(multiaddr.pop(), Some(Protocol::Quic));
+            assert_eq!(multiaddr.pop(), Some(Protocol::Udp(0)));
+            multiaddr.push(Protocol::Udp(socket_addr.port()));
+            multiaddr.push(Protocol::Quic);
+        }
+        let (new_connections, receive_connections) = mpsc::unbounded();
```

Comment:
Unbounded channels should be avoided and be replaced with bounded ones.

Reply:
The accept backlog limits how large this channel will actually become.  I could use a bounded channel and raise an internal error if its capacity is exceeded.